doc/man/radctl.8
.\" $Id$
.ds RD /usr/local/etc/raddb
.ds RL /usr/adm
.ds RP /usr/local/etc/raddb
.TH radctl 8 "March 10, 2000" "FSF"
.SH NAME
radctl - Radius daemon control interface
.SH SYNOPSIS
\fBradctl\fR -s command [options]
.br
\fBradctl\fR [\fB-d, --directory\fR DIR][\fB-S, --server\fR ID]
[\fB-u, --user\fR NAME][\fB-p, --password\fR PASS][\fB-v, --verbose\fR]
.br
\fBradctl\fR \fB-h\fR

.SH DESCRIPTION
This command is a control interface to \fBradiusd\fR daemon. It allows
administrator to query it about various aspects of its work and issue
administrative commands to it. 

\fBRadctl\fR communicates with radius using either a regular
UNIX-style signal interface or a special UDP control channel.

.SH OPTIONS
.IP "-s, --signal"
Use UNIX-style signal interface.

.IP "-d, --directory DIR"
Specifies the directory where radius dictionary files and \fBradctl.rc\fR
file reside.

.IP "-S, --server ID"
Specifies the name of the server to connect to. \fBPlease note\fR that
this is not a domain name, but rather the server ID under which it was
registered using \fI(set! rad-server-list ...)\fR statement in
\fBradctl.rc\fR file.

.IP "-u, --user NAME"
Specifies the username. If you start \fBradctl\fR without this option,
you will have to issue \fBuser\fR command from inside it.

.IP "-l, --login NAME"
An alias for \fB-u, --user\fR option.

.IP "-v, --verbose"
Enable verbose mode. In verbose mode all conversation with the
Radius server is echoed on stderr.

.IP "-h, --help"
Display short usage help and exit.

.SH COMMANDS

A command in radctl interactive mode consists of a command verb and
optional arguments. Any non-ambiguous abbreviated form of a command
verb is accepted. Command arguments, however, may not be abbreviated.

These are commands that control the operation of the \fBradctl\fR
monitor itself:

.IP help
Output short help information.

.IP "quit, bye"
Exit the program.

.IP server
Display the ID of the current radius server.

.IP "server ID"
Set current radius server.

.IP verbose
Toggle verbose mode on/off. When verbose mode is on, \fBradctl\fR
displays to stderr detailed log of its conversation with Radius
server.

.IP "login, user NAME"
Use NAME as administrative user login name.

.IP "password PASS"
Use the specified password. If `\fB.\fR' (dot) is entered instead of
plaintext password, \fBradctl\fR reads the password from /dev/tty with
character echoing turned off and generation of signals by tty special
characters disabled. This reduces the possibility of the password being
compromised.

.IP "list, list active"
Display detailed information about current radius server.

.IP "list avail"
List available radius servers. Only server ID and hostname (or IP
address) are listed.

These commands are sent directly to the server:

.IP getpid
Output the PID of the master Radius process.

.IP get-m-stat
Display detailed memory usage statistics.

.IP get-q-stat
Display detailed queue statistics. Four fields are displayed for each
queue:
.IP
Queue number
.IP
Number of pending requests
.IP
Number of processed requests still residing in the queue
.IP
Total number of requests in queue

.IP dumpdb
Tells \fBradiusd\fR to dump its user database, i.e. the parsed
contents of files \fBhints\fR, \fBhuntgroups\fR and \fBusers\fR
from the \fB\*(RD\fR directory to disk file. The filename is
\fR\*(RL/radius.parse\fB.

.SH FILES
.IP \*(RD/radctl.rc
.IP $HOME/.radctl
Initialization files.

.SH BUGS
The information in this manpage may be obsolete or incomplete. Please
refer to texinfo documentation for full information about GNU Radius
tools.

.SH AUTHORS
Sergey Poznyakoff, gray@farlep.net

.SH SEE ALSO
radiusd(8rad)


NEWS
GNU Radius NEWS -- history of user-visible changes. 2004-04-23
Copyright (C) 2002,2003,2004 Free Software Foundation, Inc.
See the end of file for copying conditions.

Please send radius bug reports to <bug-gnu-radius@gnu.org>
^L
Version 1.2.3:

* libgnuradius

This is a library of functions for creation, handling and sending
requests using RADIUS protocol.

All programs have been rewritten to link with libgnuradius. On most
sites this will mean linking against a shared library, which will
reduce the size of the executables.

* gnuradius.scm

This is a guile module allowing to use libgnuradius functions. It
supercedes radscm program, which has been removed.

* Rewrite language

** Added i18n support

** New built-in functions:

*** Functions to access internal fields of a RADIUS request.
*** Interfaces to the Radius NAS database (raddb/naslist).
*** Interfaces to DNS lookup functions.

* radiusd

** Improved checking for multiple logins. Previous versions relied entirely
on the contents of /var/log/radutmp file. Starting at this version,
radiusd offers at least two methods of checking for multiple logins:
using the traditional radutmp file and using the SQL database. New
keywords has been added to the sqlserver file that declare the SQL
queries to be used when retrieving the information about currently
active sessions.

More methods of checking will be added in future versions.

** When an unsupported authentication type is requested, radiusd first
checks if an extension Scheme module is provided that handles that
authentication type. If such module is found, it is invoked to handle
the authentication.

** System accounting can be turned off by specifying `system no;' in
`acct' section of raddb/config. 

** New configuration statement 'load-module' allows to load arbitrary
Scheme modules.

** Detailed log filenames are configurable via `detail-file-name'
statements in `auth' and `acct' sections of raddb/config.

** Support for Guile versions prior to 1.6 has been withdrawn.

* Bugfixes

** Allow to omit port numbers in `listen' statements (raddb/config), as
described in the documentation.
    


Version 1.2:

* Important notes for upgrading from earlier versions of GNU Radius

** Upgrade your dictionaries.

This version does a very strict semantic checking on the dictionary
files. Old dictionaries may produce lots of warnings and
errors, therefore you are advised to upgrade them. To do so,
run `make update-dictionary'. It will update your main 
(raddb/dictionary) and secondary dictionaries (raddb/dict/*),
preserving any changes you did to the former. 

Notice, that, starting from this version, the default main dictionary
file will not contain any attribute declarations, instead it will
merely include the necessary dictionaries. However, you may add your
declarations to it and modify it to your linking.

Any modifications to the dictionary files in raddb/dict are now
strongly discouraged, as these files will be updated by each new
release. Put your modifications to the main dictionary file only.
This file will remain unchanged during subsequent installs.

Two new statements are provided to help you customize your dictionary.
See section `** Dictionary' below.

** Update your raddb/config and raddb/rewrite

The rewrite procedures are no longer stored in a single file
`radiusd/rewrite', instead they may occupy several source file.
The rewrite procedures shipped with this version of GNU Radius
are installed in ${prefix}/share/radius/1.2/rewrite. We strongly
advise you to add the following statement to your raddb/config:

rewrite {
        load "checknas.rw";
        load "log-hook.rw";
};

You may then remove the old raddb/rewrite file, *unless* you have
modified it.

If you have modified it, follow the procedure described below:

1. Load this file, by adding the following statement to the above
`rewrite' block:

	load "rewrite";

2. Make sure to remove from this file or rename the following functions:
check_snmp_u, check_snmp_s, decimal, check_max_finger, check_unix.

** Tailor log file format.

The format of log file is now fully customizable via a special Rewrite
procedures, called "hooks" (search for logging-hook in the text below
for the detailed description). To revert to logfile format used in
versions 1.0-1.1, add the following two statements to your raddb/config,
right after `logging {' line:

       	prefix-hook "compat_log_prefix";
	suffix-hook "compat_log_suffix";

** Using Guile extensions

If you are using Guile extensions, notice that the default location
for Radius scheme modules is now ${prefix}/share/radius/1.2/scheme.
	
* New features:

** Support for Status-Server request.

** Support for Tunnel Protocol (RFC 2865)

** The main configuration file raddb/config allows to specify several
Rewrite source files to be loaded instead of the single predefined
file raddb/rewrite.

** Arbitrary Rewrite expressions may be used in A/V pairs, for example:

DEFAULT	     Auth-Type = SQL
	     Framed-IP-Address =
	           "=10.10.10.1 + %[NAS-Port-Id]/16 + %[NAS-Port-Id]"

** Test shell mode considerably improved.

** The internal attributes do not appear in detailed logs. The
special flag 'l' has been added to the dictionary syntax to
turn this feature off. Thus, for example:

ATTRIBUTE Orig-User-Name  2101	string	- [L--R-R]l

will make Orig-User-Name attribute visible in detailed log files.

** Changes to Rewrite language

*** usual dotted-quad notation may be used to represent IP numbers.

*** a number of new built-in functions is added

*** #pragma regexp allows to alter the type of regular expressions
used. The rest of line following `#pragma regexp' must consist of
whitespace-separated regexp modifiers. Each regexp modifier is one
of the following keywords:

  extended      Use extended regular expressions
  icase         Ignore case
  newline       Match-any-character operators don't match a newline.

optionally preceded by plus, indicating enabling the feature, or

radctl/radctl.c
/* This file is part of GNU RADIUS.
 * Copyright (C) 2000, Sergey Poznyakoff
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 */

#ifndef lint
static char rcsid[] = 
"$Id$";
#endif

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <sys/types.h>
#include <sys/time.h>
#include <sys/file.h>

#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <time.h>
#include <unistd.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <pwd.h>
#include <netdb.h>
#include <netinet/in.h>
#if defined(HAVE_GETOPT_LONG)
# include <getopt.h>
#endif

#include <radiusd.h>
#include <radutmp.h>
#include <radclient.h>
#define CNTL_STATE_DECL
#include <radctl.h>

#define OPTSTR "d:hLu:p:s:v?"
#ifdef HAVE_GETOPT_LONG
struct option longopt[] = {
	"directory", required_argument, 0, 'd',
	"license",   no_argument,       0, 'L',
	"server",    required_argument, 0, 's',
	"user",      required_argument, 0, 'u',
	"password",  optional_argument, 0, 'p',
	"verbose",   no_argument,       0, 'v',
	"help",      no_argument,       0, 'h',
	0
};
#else
# define longopt 0
# define getopt_long(ac,av,os,lo,li) getopt(ac,av,os)
#endif

static void usage();
static void license();
int collect_args(int argc, char **argv);
int moreinput();
void radctl();
void radctl_decode(AUTH_REQ *auth);
char * getstr_noecho(char *prompt, char *arg);
char * get_username();
void dashcmd(char *cmd, char *arg);
void bye();
void source_rc();

int             verbose;
char            *hostname;
char            *user;
char            *password;
char		vector[AUTH_VECTOR_LEN];

static char     *prompt = "> ";
static char     *filename;
static int      line_no;
static char     input_buffer[216];
static char     *input_ptr;
RADCLIENT       *radclient;

int
main(argc, argv)
	int argc;
	char **argv;
{
	int c;
	
	app_setup();
	initlog(argv[0]);

	while ((c = getopt_long(argc, argv, OPTSTR, longopt, NULL)) != EOF) 
	        switch(c) {
		case 'd':
			radius_dir = optarg;
			break;
		case 's':
			hostname = optarg;
			break;
		case 'u':
			user = optarg;
			break;
		case 'p':
			password = optarg;
			break;
		case 'v':
			verbose++;
			break;
		case 'L':
			license();
			exit(0);
		case 'h':
		case '?':
		default:
			usage();
			/*NOTREACHED*/
	}

	if (!user && (user = get_username()) == NULL) 
		return 1;
	password = getstr_noecho(_("Password"), password);
		       
	radpath_init();
	if (dict_init(NULL) < 0) {
		return -1;
	}
	radclient = radclient_init(radius_dir);
	if (!radclient)
		return -1;
	
	radclient_delete_server(radclient, NULL);
	if (hostname && radclient_find_server(radclient, hostname) == 0) {
		radlog(L_ERR, _("unknown server: %s"), hostname);
		hostname = NULL;
	}
	if (!hostname)
		hostname = radclient->server->name;
	radclient_add_server(radclient, hostname);
	
	
	if (collect_args(argc - optind, argv + optind)) {
		filename = _("<command line>");
		line_no = 0;
		radctl();
	} else {
		filename = _("<teletype>");
		line_no = 0;
		while (moreinput(stdin))
			radctl();
		bye();
	}
	return 0;
}

void
change_hostname(new_name)
	char *new_name;
{
	if (radclient_find_server(radclient, new_name) == 0) {
		radlog(L_ERR, _("unknown hostname: %s"), new_name);
		return;
	}
	radclient_delete_server(radclient, NULL);
	radclient_add_server(radclient, new_name);
}


int
collect_args(argc, argv)
	int argc;
	char **argv;
{
	char *p;
	int len;
	
	p = input_buffer;
	while (argc--) {
		len = strlen(*argv);
		if (p + len + 1 >= input_buffer + sizeof input_buffer) {
			radlog(L_ERR, _("argument command string too long: truncated"));
			break;
		}
		strncpy(p, *argv, len);
		p += len;
		*p++ = ' ';
		++argv;
	}
	*p = 0;
	input_ptr = input_buffer;
	return input_ptr[0];
}

void
bye()
{
	printf(_("bye\n"));
	exit(0);
}

void
source_rc(dir, file)
	char *dir;
	char *file;
{
	FILE *fp;
	char *path = mkfilename(dir, file);

	if ((fp = fopen(path, "r")) == NULL) {
		if (errno != ENOENT)
			radlog(L_ERR|L_PERROR,
			       _("can't open `%s' for reading"),
			       path);
		efree(path);
		return;
	}
	filename = path;
	line_no = 0;
	while (moreinput(fp))
		radctl();
	fclose(fp);
	efree(path);
}

int
moreinput(fp)
	FILE *fp;
{
	if (isatty(fileno(fp)))
		printf("%s", prompt);
	input_ptr = fgets(input_buffer, sizeof(input_buffer), fp);
	line_no++;
	return input_ptr != NULL;
}

void
radctl()
{
	char *statep, *argp;
	VALUE_PAIR *pairlist;
	AUTH_REQ   *req;
	
	if (!input_ptr)
		return;
	
	while (*input_ptr && isspace(*input_ptr))
		input_ptr++;
	if (*input_ptr == 0 || *input_ptr == '#')
		return;
	
	statep = input_ptr;
	while (*input_ptr && !isspace(*input_ptr))
		input_ptr++;
	if (*input_ptr) {
		*input_ptr++ = 0;
		while (*input_ptr && isspace(*input_ptr))
			input_ptr++;
		argp = input_ptr;
		while (*input_ptr && !isspace(*input_ptr))
			input_ptr++;
		if (*input_ptr)
			*input_ptr++ = 0;
	} else
		argp = NULL;

	if (argp && *argp == 0)
		argp = NULL;
	
	if (statep[0] == '-') {
		dashcmd(statep, argp);
		return;
	}

	pairlist = create_pair(DA_USER_NAME, strlen(user), user, 0);
	pairadd(&pairlist,
		create_pair(DA_PASSWORD, strlen(password), password, 0));
	pairadd(&pairlist,
		create_pair(DA_STATE, strlen(statep), statep, 0));
	if (argp)
		pairadd(&pairlist,
			create_pair(DA_CLASS, strlen(argp), argp, 0));

	req = radclient_send(radclient, 2, PW_AUTHENTICATION_REQUEST, pairlist);
	pairfree(pairlist);
	if (req) {
		radctl_decode(req);
		authfree(req);
	} else
		printf(_("no answer from %s\n"),
			 ip_hostname(radclient->first_server->addr));
}

void
radctl_decode(auth)
	AUTH_REQ *auth;
{
	VALUE_PAIR *pair;
	
	if (auth->code == PW_AUTHENTICATION_ACK)
		printf(_("OK\n"));
	else
		printf(_("ERROR\n"));

	for (pair = auth->request; pair; pair = pair->next) {
		if (pair->attribute == DA_REPLY_MESSAGE) 
			printf("%s", pair->strvalue);
	}
	printf("\n");
}

enum dash_command {
	CMD_HELP,
	CMD_QUIT,
	CMD_SERVER,
	CMD_VERBOSE,
	CMD_USER,
	CMD_PASSWORD,
};

struct keyword dash_kwd[] = {
	"-help",     CMD_HELP,
	"-?",        CMD_HELP,
	"-quit",     CMD_QUIT,
	"-q",        CMD_QUIT,
	"-s",        CMD_SERVER,
	"-server",   CMD_SERVER,
	"-verbose",  CMD_VERBOSE,
	"-v",        CMD_VERBOSE,
	"-user",     CMD_USER,
	"-u",        CMD_USER,
	"-password", CMD_PASSWORD,
	"-p",        CMD_PASSWORD,
	0
};

char *cmd_usage_str = N_(
"commands:\n"
"    -help or -?         display this help\n"
"    -q(uit)             quit the program\n"
"    -s(server)          change RADIUS server\n"
"    -verbose            toggle verbose mode\n"
"    -u(ser)             set user name\n"
"    -p(assword)         set user password\n");


void
dashcmd(s, arg)
	char *s;
	char *arg;
{
	extern int radclient_debug;
	
	if (s[1] == '-')
		s++;
	switch (xlat_keyword(dash_kwd, s, -1)) {
	case CMD_HELP:
		printf("%s", _(cmd_usage_str));
		break;
		
	case CMD_QUIT:
		bye();
		
	case CMD_SERVER:
		change_hostname(arg);
		break;
		
	case CMD_VERBOSE:
		verbose = !verbose;
		printf(_("radctl is now %s\n"),
		       verbose ? _("verbose") : _("silent"));
		radclient_debug = verbose;
		break;
		
	case CMD_USER:
		if (arg)
			user = estrdup(arg);
		else {
			char *p;
			
			p = get_username();
			if (!p) 
				printf(_("not changed\n"));
			else
				user = p;
		}
		break;
		
	case CMD_PASSWORD:
		password = getstr_noecho(_("Password"), arg);
		break;

	default:
		radlog(L_ERR, _("%s:%d: unknown command: %s"),
		       filename, line_no, s);
	}
}


static char usage_str[] =
#ifdef HAVE_GETOPT_LONG
N_(
"usage: radctl [options] [commands]\n"
"Options are:\n"
"    -d, --directory DIR          specify radius database directory\n"
"    -L, --license                display GNU license and exit\n"
"    -s, --server SERVERNAME      set RADIUS server name\n"
"    -u, --user USERNAME          set user name\n"
"    -p, --password TEXT          set user password\n"
"    -v, --verbose                be verbose on output\n"
"    -?, --help                   display this help\n")
#else
N_("usage: radctl [options] [commands]\n"
"Options are:\n"
"    -d DIR        specify radius database directory\n"
"    -L            display GNU license and exit\n"
"    -s SERVERNAME set RADIUS server name\n"
"    -u USERNAME   set user name\n"
"    -p TEXT       set user password\n"
"    -v            be verbose on output\n"
"    -?            display this help\n")
#endif
;

void
usage()
{
	printf(_(usage_str));
	exit(0);
}

char *
getstr_noecho(prompt, arg)
	char *prompt;
	char *arg;
{
	if (arg == NULL || strcmp(arg, ":") == 0) {
		printf("%s", prompt);fflush(stdout);
		arg = getpass(":");
	} 
	return estrdup(arg);
}


char *
get_username()
{
	char buffer[RUT_NAMESIZE+1];
	int len;
	
	printf(_("Username: "));
	fgets(buffer, sizeof(buffer), stdin);
	len = strlen(buffer);
	if (len <= 1)
		return NULL;
	if (buffer[len-1] == '\n')
		buffer[len-1] = 0;
	return estrdup(buffer);
}

char license_text[] =
"   This program is free software; you can redistribute it and/or modify\n"
"   it under the terms of the GNU General Public License as published by\n"
"   the Free Software Foundation; either version 2, or (at your option)\n"
"   any later version.\n"
"\n"
"   This program is distributed in the hope that it will be useful,\n"
"   but WITHOUT ANY WARRANTY; without even the implied warranty of\n"
"   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"
"   GNU General Public License for more details.\n"
"\n"
"   You should have received a copy of the GNU General Public License\n"
"   along with this program; if not, write to the Free Software\n"
"   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n";

void
license()
{
	printf("%s: Copyright 1999,2000 Sergey Poznyakoff\n", progname);
	printf("\nThis program is part of GNU Radius\n");
	printf("%s", license_text);
}
